common/logisticspipes/utils/string/StringUtils.java

```diff
@@ -187,10 +189,11 @@ public static String insertThousandsSeparators(String source) {
 		for (String paragraph : paragraphs) {
 			boolean header = paragraph.indexOf("##") > -1;
 			if (header) paragraph = paragraph.replace("##", "");
-			if (paragraph.charAt(paragraph.length() - 1) != '\n') paragraph += '\n';
+			if (paragraph.length() != 0 && paragraph.charAt(paragraph.length() - 1) != '\n') paragraph += (TextFormatting.WHITE + "\n");
```

Comment:
@Bernasss12 if the `paragraph` is null, then `paragraph.length()` will generate NPE. 

Suggested solution:

```
if (paragraph != null && 
    paragraph.length() != 0 && 
    paragraph.charAt(paragraph.length() - 1) != '\n') {
        paragraph += (TextFormatting.WHITE + "\n");
}
```

Reply:
I don't know how a null paragraph would come from a `split()` but I guess it's better to be safe than sorry

Reply:
Honestly I didn't checked the source of the string array, but later maybe you refactor this code and feed paragraphs from different sources (like JSON parser or NBT parser or anything else) and at that point you will be thankful for this.

Reply:
All of this part of the code will drastically change in the near future so this was just a quick fix so people can start populating the book itself, but yes thank you for that catch

Reply:
I disagree. `paragraph` isn't null here as it's String methods are used the two lines above that line. Adding the null check here adds confusion in any case. Assuming that the `paragraphs` array does not contain null values is absolutely valid and any refactoring should check null before inserting into the array then.

The general rule should be to not use null and mark all other code with `@Nullable` from the `javax.annotation` package. Just to clarify this bit. I understand that the code will change.